### PR TITLE
dense_norms

### PR DESCRIPTION
**Pull request by MrRobot2211**
This implements the noms in data.norms

Tests are not present in QuTiP so they are created here for the sake of speeding things up. But we really should have the tests on dev.major and the inherit them here.

This is ready to merge after extra-functions.

### CODE REVIEW

**Comment by hodgestar** [comment]
@MrRobot2211 Could you merge the main branch into this one now that extra functions has been merged? At the moment the diff is large and contains mostly stuff that is already reviewed and merged into main.

File: src/qutip_cupy/dense_functions.py [hunk 1]
```diff
@@ -29,3 +29,23 @@ def trace_cupydense(cp_arr):
...
+        raise ValueError("L2 norm is only defined on vectors")
+    return frobenius_cupydense(cp_arr)
+
+
+def max_cupydense(cp_arr):
+    return cp.max(cp.abs(cp_arr._cp)).item()
+
+
+def one_cupydense(cp_arr):
+
```
Comment (hodgestar) [inline src/qutip_cupy/dense_functions.py:50]:
```suggestion
```

File: src/qutip_cupy/__init__.py [hunk 2]
```diff
@@ -58,6 +58,12 @@
 data.trace.add_specialisations([(CuPyDense, cdf.trace_cupydense)])
 data.reshape.add_specialisations([(CuPyDense, CuPyDense, cdf.reshape_cupydense)])
 
+# the whole norm module is imported in QuTiP
+# so the dispatching is peculiar
+data.norm.l2.add_specialisations([(CuPyDense, cdf.l2_cupydense)])
+data.norm.frobenius.add_specialisations([(CuPyDense, cdf.frobenius_cupydense)])
+data.norm.max.add_specialisations([(CuPyDense, cdf.max_cupydense)])
+data.norm.one.add_specialisations([(CuPyDense, cdf.one_cupydense)])
```
Comment (hodgestar) [inline src/qutip_cupy/__init__.py:66]:
```suggestion
data.norm.l2.add_specialisations([(CuPyDense, cdf.l2_cupydense)])
data.norm.frobenius.add_specialisations([(CuPyDense, cdf.frobenius_cupydense)])
data.norm.max.add_specialisations([(CuPyDense, cdf.max_cupydense)])
data.norm.one.add_specialisations([(CuPyDense, cdf.one_cupydense)])

```
Putting these into a separate paragraph and removing the comment about the dispatching being weird. This is not weird -- namespacing is a core Python feature!

File: tests/test_mathematics.py [hunk 3]
```diff
@@ -86,3 +87,88 @@ class TestTranspose(test_tools.TestTranspose):
...
+        (pytest.param((100, 100), id="100"),),
+        (pytest.param((23, 30), id="23_30"),),
+        (pytest.param((15, 10), id="15_10"),),
+    ]
+
+    specialisations = [
+        pytest.param(cdf.l2_cupydense, CuPyDense, float),
+    ]
+
+    # Trace actually does have bad shape, so we put that in too.
```
Comment (hodgestar) [inline tests/test_mathematics.py:131]:
```suggestion
    # l2 norm actually does have bad shape, so we put that in too.
```

**Review by hodgestar — CHANGES_REQUESTED** [review]
Left some small suggestions, but otherwise looks good.

**Comment by MrRobot2211** [comment]
Ready to merge

**Review by hodgestar — APPROVED** [review]
Approved!

**Pull request merged**